embrace-gradle-plugin/src/main/java/io/embrace/android/gradle/plugin/tasks/ndk/CompressSharedObjectFilesTask.kt

```diff
@@ -0,0 +1,58 @@
+package io.embrace.android.gradle.plugin.tasks.ndk
+
+import io.embrace.android.gradle.plugin.tasks.EmbraceTaskImpl
+import io.embrace.android.gradle.plugin.util.compression.ZstdFileCompressor
+import org.gradle.api.file.DirectoryProperty
+import org.gradle.api.model.ObjectFactory
+import org.gradle.api.tasks.InputDirectory
+import org.gradle.api.tasks.OutputDirectory
+import org.gradle.api.tasks.SkipWhenEmpty
+import org.gradle.api.tasks.TaskAction
+import java.io.File
+import javax.inject.Inject
+
+/**
+ * Given a directory containing architecture directories (arm64-v8a, armeabi-v7a, etc.) with shared object files (.so files),
+ * this task compresses each shared object file using Zstd compression and stores the compressed files in the output directory,
+ * preserving the architecture directory structure.
+ */
+abstract class CompressSharedObjectFilesTask @Inject constructor(
+    objectFactory: ObjectFactory,
+) : EmbraceTaskImpl(objectFactory) {
+
+    private val compressor = ZstdFileCompressor()
+
+    @get:InputDirectory
+    @get:SkipWhenEmpty
+    val architecturesDirectory: DirectoryProperty = objectFactory.directoryProperty()
+
+    @get:OutputDirectory
+    val compressedSharedObjectFilesDirectory: DirectoryProperty = objectFactory.directoryProperty()
+
+    @TaskAction
+    fun onRun() {
+        architecturesDirectory.get().asFile
+            .listFiles()
+            ?.filter { it.isDirectory && it.listFiles()?.isNotEmpty() == true }
```

Comment:
nit: if you use ` ?: emptyList()` you can avoid the null coalescening operator on the rest of the functional chain